### PR TITLE
[ADP-3198] Remove record representation of erafun

### PR DESCRIPTION
**Pull request by paolino**
- [x] simplify EraFun creation and destruction by switching from real record to pattern record
- [x] add a bidirectional pattern for AllEraValue values

### CODE REVIEW

File: lib/read/lib/Cardano/Wallet/Read/Eras/EraFun.hs [hunk 1]
```diff
@@ -3,12 +3,13 @@
 {-# LANGUAGE InstanceSigs #-}
 {-# LANGUAGE MultiParamTypeClasses #-}
 {-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE QuantifiedConstraints #-}
```
Comment (HeinrichApfelmus) [inline lib/read/lib/Cardano/Wallet/Read/Eras/EraFun.hs]:
```suggestion
```

Not used here.

File: lib/read/lib/Cardano/Wallet/Read/Eras/EraFun.hs [hunk 2]
```diff
@@ -3,12 +3,13 @@
 {-# LANGUAGE InstanceSigs #-}
 {-# LANGUAGE MultiParamTypeClasses #-}
```
Comment (HeinrichApfelmus) [inline lib/read/lib/Cardano/Wallet/Read/Eras/EraFun.hs]:
```suggestion
```

Not used anymore.

File: lib/read/lib/Cardano/Wallet/Read/Eras/EraFun.hs [hunk 3]
```diff
@@ -194,42 +213,58 @@ newtype EraFunK src ft = EraFunK {fromEraFunK :: EraFun src (K ft)}
...
+        ( EraFun
+                { byronFun = (mkConst -> byronVal)
+                , shelleyFun = (mkConst -> shelleyVal)
+                , allegraFun = (mkConst -> allegraVal)
+                , maryFun = (mkConst -> maryVal)
+                , alonzoFun = (mkConst -> alonzoVal)
+                , babbageFun = (mkConst -> babbageVal)
+                , conwayFun = (mkConst -> conwayVal)
+                }
+            )
```
Comment (HeinrichApfelmus) [inline lib/read/lib/Cardano/Wallet/Read/Eras/EraFun.hs:277]:
```suggestion
  where
    AllEraValueP
        byronVal
        shelleyVal
        allegraVal
        maryVal
        alonzoVal
        babbageVal
        conwayVal
        = AllEraValue
            ( EraFun
                { byronFun = const byronVal
                , shelleyFun = const shelleyVal
                , allegraFun = const allegraVal
                , maryFun = const maryVal
                , alonzoFun = const alonzoVal
                , babbageFun = const babbageVal
                , conwayFun = const conwayVal
                }
            )
```

The comment promises "construct/deconstruct", so we should make the pattern bidirectional.

Either that, or change the comment to
```
-- | A pattern to deconstruct an 'AllEraValue'
```

**Review by HeinrichApfelmus — APPROVED** [review]
Excellent, thank you! 😊 I would suggest to remove more language extensions and to make the pattern bidirectional, though.

File: lib/read/lib/Cardano/Wallet/Read/Eras/EraFun.hs (hunk 3 above)
Comment (paolino) [reply lib/read/lib/Cardano/Wallet/Read/Eras/EraFun.hs:277]: I think it's bidirectional, and I cannot make your suggestion to compile

Comment (paolino) [reply lib/read/lib/Cardano/Wallet/Read/Eras/EraFun.hs:277]: no it's not bidirectional

Comment (paolino) [reply lib/read/lib/Cardano/Wallet/Read/Eras/EraFun.hs:277]:
```haskell
-- | A pattern to construct/deconstruct an 'AllEraValue'
pattern AllEraValueP
    :: f ByronEra
    -> f ShelleyEra
    -> f AllegraEra
    -> f MaryEra
    -> f AlonzoEra
    -> f BabbageEra
    -> f ConwayEra
    -> AllEraValue f
pattern AllEraValueP
    { byronVal
    , shelleyVal
    , allegraVal
    , maryVal
    , alonzoVal
    , babbageVal
    , conwayVal
    } <-
    AllEraValue
        ( EraFun
                { byronFun = mkConst -> byronVal
                , shelleyFun = mkConst -> shelleyVal
                , allegraFun = mkConst -> allegraVal
                , maryFun = mkConst -> maryVal
                , alonzoFun = mkConst -> alonzoVal
                , babbageFun = mkConst -> babbageVal
                , conwayFun = mkConst -> conwayVal
                }
            )
    where
        AllEraValueP
            byronVal'
            shelleyVal'
            allegraVal'
            maryVal'
            alonzoVal'
            babbageVal'
            conwayVal' =
                AllEraValue
                    $ EraFun
                        { byronFun = const byronVal'
                        , shelleyFun = const shelleyVal'
                        , allegraFun = const allegraVal'
                        , maryFun = const maryVal'
                        , alonzoFun = const alonzoVal'
                        , babbageFun = const babbageVal'
                        , conwayFun = const conwayVal'
                        }
```
is working.
I think we need this weird syntax as we are pattern matching on a part of the value so we need `get` and `set`

**Pull request merged**